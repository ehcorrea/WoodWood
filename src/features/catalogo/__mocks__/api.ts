import { mocked } from 'jest-mock';
import * as api from '../api/api';

jest.mock('../api/api');

export const mockedApi = mocked(api);
