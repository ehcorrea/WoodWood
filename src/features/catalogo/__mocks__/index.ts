export * from './api';
export * from './produtos';
