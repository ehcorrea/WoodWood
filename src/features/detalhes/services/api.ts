import { client } from '@/shared/services/client';
import { Produto } from '@/shared/types';

export type DetalhesProdutoArgs = {
  id: number;
};

export async function detalhesProduto({
  id,
}: DetalhesProdutoArgs): Promise<Produto> {
  const { data } = await client.get(`/products/${id}`);
  return data;
}
