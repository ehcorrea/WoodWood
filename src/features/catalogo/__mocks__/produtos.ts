import { Produto } from '../api/types';

export const MOCKED_PRODUTOS: Produto[] = [
  {
    category: 'mocked-category',
    description: 'mocked-description',
    id: 0,
    image: 'mocked-image',
    price: 100,
    rating: {
      count: 10,
      rate: 10,
    },
    title: 'mocked-title',
  },
];
