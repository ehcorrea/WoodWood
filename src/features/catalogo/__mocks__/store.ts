import { mockedStore } from '@/shared/__mocks__';

import { catalogoStore, CatalogoStore } from '../store';

export const mockedCatalagoStore = (storeValues?: Partial<CatalogoStore>) =>
  mockedStore({
    store: catalogoStore,
    storeValues: { ...catalogoStore.getState(), ...storeValues },
  });
