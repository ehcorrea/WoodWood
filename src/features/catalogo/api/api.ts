import { client } from '../../../shared/services/client';

import { Produto } from './types';

export async function listarProdutos(): Promise<Produto[]> {
  const { data } = await client.get('/products');
  return data;
}
