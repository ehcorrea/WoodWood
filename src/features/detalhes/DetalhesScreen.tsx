import { ScrollView, View } from 'react-native';
import { RouteProp, useRoute } from '@react-navigation/native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';

import { NavigationParamList } from '@/navigators/RootNavigator.types';
import { Produto } from '@/services/api';
import { Spacing, Text } from '@/shared/components';
import { carrinhoStore } from '@/shared/stores';

import { useDetalhesProduto } from './hooks';

import * as S from './DetalhesScreen.styles';

export type DetalheScreen = RouteProp<NavigationParamList, 'DetalhesScreen'>;

export default function DetalhesScreen() {
  const route = useRoute<DetalheScreen>();
  const { bottom } = useSafeAreaInsets();
  const { adicionarProduto } = carrinhoStore();
  const { produto } = useDetalhesProduto(route.params);

  const handleAdicionarProduto = (produtoParaAdicionar: Produto) => {
    const { id, image, price, title } = produtoParaAdicionar;
    adicionarProduto({
      id,
      image,
      price,
      title,
    });
  };

  return (
    <S.Wrapper>
      <ScrollView
        bounces={false}
        showsVerticalScrollIndicator={false}
        contentContainerStyle={{ paddingBottom: bottom + 85 }}
      >
        <S.Image source={{ uri: produto?.image }} resizeMode="contain" />
        <Spacing y={10} />
        <View>
          <Text.Title size="xlarge">{produto?.title}</Text.Title>
          <Spacing y={10} />
          <Text.Subtitle>R$ {produto?.price}</Text.Subtitle>
          <Spacing y={10} />
          <Text>R$ {produto?.description}</Text>
          <Spacing y={10} />
        </View>
      </ScrollView>
      {produto && (
        <S.ButtonAdicionar
          bottom={bottom}
          onPress={() => handleAdicionarProduto(produto)}
        >
          <Text.Subtitle palette="white">Adicionar ao carrinho</Text.Subtitle>
        </S.ButtonAdicionar>
      )}
    </S.Wrapper>
  );
}
