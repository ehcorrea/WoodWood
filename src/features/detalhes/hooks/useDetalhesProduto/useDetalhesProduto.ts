import { useEffect } from 'react';
import { Alert } from 'react-native';
import { useQuery } from '@tanstack/react-query';

import { detalhesProduto } from '@/services/api';

export type UseDetalhesProdutoArgs = {
  id: number;
};

export function useDetalhesProduto({ id }: UseDetalhesProdutoArgs) {
  const { data, isError, refetch, isLoading } = useQuery({
    queryKey: ['detalhes-produto', id],
    queryFn: () => detalhesProduto({ id }),
  });

  useEffect(() => {
    if (isError) {
      Alert.alert('Error', 'Tentar Novamente?', [
        { text: 'Sim', onPress: () => refetch() },
        { text: 'NÃ£o' },
      ]);
    }
  }, [isError]);

  return { produto: data, isLoading };
}
