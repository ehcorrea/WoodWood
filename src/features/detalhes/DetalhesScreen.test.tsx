import { fireEvent } from '@testing-library/react-native';
import * as navigation from '@react-navigation/native';

import { render } from '@/test/utils';
import { criarMockProduto, mockedCarrinhoStore } from '@/test/__mocks__/';
import { Produto } from '@/services/api';

import * as useDetalhesProduto from './hooks/useDetalhesProduto/useDetalhesProduto';

import DetalhesScreen from './DetalhesScreen';

type SetupArgs = {
  id: number;
  produto?: Produto;
  isLoading?: boolean;
};

const adicionarProduto = jest.fn();

const setup = ({ id, produto, isLoading = false }: SetupArgs) => {
  mockedCarrinhoStore({ adicionarProduto });
  jest.spyOn(navigation, 'useRoute').mockReturnValueOnce({
    params: { id },
    key: 'detalhes-screen',
    name: 'DetalhesScreen',
  });
  jest
    .spyOn(useDetalhesProduto, 'useDetalhesProduto')
    .mockReturnValueOnce({ produto: produto, isLoading });
  return render(<DetalhesScreen />);
};

describe('<DetalhesScreen/>', () => {
  describe('quando renderizado', () => {
    test('e em loading', () => {
      const container = setup({ id: 1, isLoading: true });
      expect(container.getByTestId('loading')).toBeTruthy();
      expect(container.queryByText('Adicionar ao carrinho')).toBeFalsy();
    });
    describe('e fora de loading', () => {
      describe('e pressionado "Adicionar ao carrinho"', () => {
        test('com produto', () => {
          const produto = criarMockProduto();
          const container = setup({ id: produto.id, produto });
          expect(container.queryByTestId('loading')).toBeFalsy();
          const adicionarAoCarrinho = container.getByText(
            'Adicionar ao carrinho'
          );
          expect(useDetalhesProduto.useDetalhesProduto).toHaveBeenCalledWith({
            id: produto.id,
          });
          fireEvent.press(adicionarAoCarrinho);
          expect(adicionarProduto).toHaveBeenCalledWith({
            id: produto.id,
            image: produto.image,
            price: produto.price,
            title: produto.title,
          });
        });
        test('sem produto', () => {
          const produto = criarMockProduto();
          const container = setup({ id: produto.id });
          expect(useDetalhesProduto.useDetalhesProduto).toHaveBeenCalledWith({
            id: 0,
          });
          expect(container.queryByText('Adicionar ao carrinho')).toBeFalsy();
        });
      });
    });
  });
});
