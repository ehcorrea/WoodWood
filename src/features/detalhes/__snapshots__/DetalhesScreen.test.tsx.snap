// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetalhesScreen/> quando renderizado e em loading 1`] = `
<ActivityIndicator
  size="large"
  testID="loading"
/>
`;

exports[`<DetalhesScreen/> quando renderizado e fora de loading e pressionado "Adicionar ao carrinho" com produto 1`] = `
<View
  style={
    [
      {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": "5%",
        "paddingLeft": "5%",
        "paddingRight": "5%",
        "paddingTop": "5%",
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    bounces={false}
    contentContainerStyle={
      {
        "paddingBottom": 95,
      }
    }
    showsVerticalScrollIndicator={false}
  >
    <View>
      <Image
        resizeMode="contain"
        source={
          {
            "uri": "mocked-image",
          }
        }
        style={
          [
            {
              "aspectRatio": 1,
              "width": "100%",
            },
            undefined,
          ]
        }
      />
      <View
        style={
          [
            {
              "marginHorizontal": 0,
              "marginVertical": 10,
            },
            undefined,
          ]
        }
        testID="spacing"
        x={0}
        y={10}
      />
      <View>
        <Text
          size="xlarge"
          style={
            [
              {
                "color": "#1A1D26",
                "fontFamily": "SofiaPro-Bold",
                "fontSize": 34.5,
                "lineHeight": 43,
              },
              undefined,
            ]
          }
          weight="bold"
        >
          mocked-title
        </Text>
        <View
          style={
            [
              {
                "marginHorizontal": 0,
                "marginVertical": 10,
              },
              undefined,
            ]
          }
          testID="spacing"
          x={0}
          y={10}
        />
        <Text
          size="large"
          style={
            [
              {
                "color": "#1A1D26",
                "fontFamily": "SofiaPro-SemiBold",
                "fontSize": 25.5,
                "lineHeight": 28.5,
              },
              undefined,
            ]
          }
          weight="semibold"
        >
          R$ 
          100
        </Text>
        <View
          style={
            [
              {
                "marginHorizontal": 0,
                "marginVertical": 10,
              },
              undefined,
            ]
          }
          testID="spacing"
          x={0}
          y={10}
        />
        <Text
          style={
            [
              {
                "color": "#1A1D26",
                "fontFamily": "SofiaPro-Regular",
                "fontSize": 21.5,
                "lineHeight": 25.5,
              },
              undefined,
            ]
          }
        >
          R$ 
          mocked-description
        </Text>
        <View
          style={
            [
              {
                "marginHorizontal": 0,
                "marginVertical": 10,
              },
              undefined,
            ]
          }
          testID="spacing"
          x={0}
          y={10}
        />
      </View>
    </View>
  </RCTScrollView>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#EA5B0C",
        "borderBottomLeftRadius": 30,
        "borderBottomRightRadius": 30,
        "borderTopLeftRadius": 30,
        "borderTopRightRadius": 30,
        "bottom": 10,
        "height": 50,
        "justifyContent": "center",
        "opacity": 1,
        "position": "absolute",
        "width": "90%",
      }
    }
  >
    <Text
      palette="white"
      size="large"
      style={
        [
          {
            "color": "#FFFFFF",
            "fontFamily": "SofiaPro-SemiBold",
            "fontSize": 25.5,
            "lineHeight": 28.5,
          },
          undefined,
        ]
      }
      weight="semibold"
    >
      Adicionar ao carrinho
    </Text>
  </View>
</View>
`;

exports[`<DetalhesScreen/> quando renderizado e fora de loading e pressionado "Adicionar ao carrinho" sem produto 1`] = `
<View
  style={
    [
      {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": "5%",
        "paddingLeft": "5%",
        "paddingRight": "5%",
        "paddingTop": "5%",
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    bounces={false}
    contentContainerStyle={
      {
        "paddingBottom": 95,
      }
    }
    showsVerticalScrollIndicator={false}
  >
    <View>
      <Image
        resizeMode="contain"
        source={
          {
            "uri": undefined,
          }
        }
        style={
          [
            {
              "aspectRatio": 1,
              "width": "100%",
            },
            undefined,
          ]
        }
      />
      <View
        style={
          [
            {
              "marginHorizontal": 0,
              "marginVertical": 10,
            },
            undefined,
          ]
        }
        testID="spacing"
        x={0}
        y={10}
      />
      <View>
        <Text
          size="xlarge"
          style={
            [
              {
                "color": "#1A1D26",
                "fontFamily": "SofiaPro-Bold",
                "fontSize": 34.5,
                "lineHeight": 43,
              },
              undefined,
            ]
          }
          weight="bold"
        />
        <View
          style={
            [
              {
                "marginHorizontal": 0,
                "marginVertical": 10,
              },
              undefined,
            ]
          }
          testID="spacing"
          x={0}
          y={10}
        />
        <Text
          size="large"
          style={
            [
              {
                "color": "#1A1D26",
                "fontFamily": "SofiaPro-SemiBold",
                "fontSize": 25.5,
                "lineHeight": 28.5,
              },
              undefined,
            ]
          }
          weight="semibold"
        >
          R$ 
        </Text>
        <View
          style={
            [
              {
                "marginHorizontal": 0,
                "marginVertical": 10,
              },
              undefined,
            ]
          }
          testID="spacing"
          x={0}
          y={10}
        />
        <Text
          style={
            [
              {
                "color": "#1A1D26",
                "fontFamily": "SofiaPro-Regular",
                "fontSize": 21.5,
                "lineHeight": 25.5,
              },
              undefined,
            ]
          }
        >
          R$ 
        </Text>
        <View
          style={
            [
              {
                "marginHorizontal": 0,
                "marginVertical": 10,
              },
              undefined,
            ]
          }
          testID="spacing"
          x={0}
          y={10}
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
